src/include/miopen/fin/fin_interface.hpp

```diff
@@ -0,0 +1,247 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#pragma once
+
+#include <cstdint>
+#include <string>
+#include <vector>
+
+#include <miopen/conv_algo_name.hpp>
+#include <miopen/conv_solution.hpp>
+#include <miopen/miopen.h>
+#include <miopen/mlo_internal.hpp>
+
+namespace miopen {
+
+struct AnyInvokeParams;
+struct ExecutionContext;
+
+namespace conv {
+struct ProblemDescription;
+} // namespace conv
+
+namespace batchnorm {
+struct ProblemDescription;
+} // namespace batchnorm
+
+namespace solver {
+struct SolverBase;
+} // namespace solver
+
+namespace fin {
+
+// Base classes for solvers.
+class Solver
+{
+public:
+    // GetId(), IsDynamic() and IsTunable() throw miopenStatusNotInitialized if the solver is not
+    // valid.
+
+    // Returns false if the solver could not be found by its name.
+    bool IsValid() const;
+
+    uint64_t GetId() const;
+    // Returns the name even if the solver is not valid (returns the requested name).
+    const std::string& GetName() const;
+
+    bool IsTunable() const;
+    bool IsDynamic() const;
+
+protected:
+    Solver(const miopen::solver::SolverBase* solver_base, uint64_t solver_id);
+    Solver(const std::string& requested_name);
+
+    const miopen::solver::SolverBase* const sbase = nullptr;
+    const std::string rname;
+    uint64_t id;
+
+    friend class FinInterface;
+};
+
+template <class Context, class Problem>
+class SolverMixin : public Solver
+{
+public:
+    // All the methods throw miopenStatusNotInitialized if the solver is not valid.
+
+    bool IsApplicable(const Context& ctx, const Problem& problem) const;
+
+    size_t GetWorkspaceSize(const Context& ctx, const Problem& problem) const;
+
+    miopen::solver::ConvSolution FindSolution(const Context& ctx,
+                                              const Problem& problem,
+                                              miopen::PerformanceDb& db,
+                                              const miopen::AnyInvokeParams& invoke_ctx,
+                                              const std::string& perf_cfg = "") const;
+
+    std::vector<miopen::solver::ConvSolution> GetAllSolutions(const Context& ctx,
+                                                              const Problem& problem) const;
+
+    std::string
+    GetPerfCfgParams(const Context& ctx, const Problem& problem, const PerformanceDb& db) const;
+
+    bool
+    TestPerfCfgParams(const Context& ctx, const Problem& problem, const std::string& params) const;
+
+protected:
+    using Solver::Solver;
+};
+
+// Convolution solver
+class ConvSolver : public SolverMixin<miopen::ExecutionContext, miopen::conv::ProblemDescription>
+{
+public:
+    std::string GetAlgo(miopen::conv::Direction dir) const;
+
+protected:
+    ConvSolver(const miopen::solver::SolverBase* solver_base,
+               uint64_t solver_id,
```

Comment:
I would suggest using `solver::Id` instead of numeric ids. It would be hassle free to create this from either number or string.